### ISSUE TITLE
Implement cp

### ISSUE THREAD

**Issue by GrayJack** [GrayJack/coreutils#15]

**Comment by kegesch** [comment]
I will start with this.

**Comment by GrayJack** [comment]
Nice, have fun :smile:

**Comment by kegesch** [comment]
@GrayJack i ran into a problem here. Maybe you know how to solve it. `cp -p -R symlink` should copy a symlink and preserve its metadata. But apparently there is no way in Rust to copy a symlink. You can only create one. But also you can not set a file's metadata, at least i could not find any way to do so. Do you have an idea?

**Comment by GrayJack** [comment]
I have to look into that, never done it

**Comment by GrayJack** [comment]
How do you preserve the metadata on the other types of file?

**Comment by kegesch** [comment]
`fs::copy()` copies a file with metadata. And to not preserve it i  use `io::cop()`. For `dir`s  i have the same problem as with symlinks. There is function in `fs` to set permissions but i found nothing about modified, creation date etc.

**Comment by GrayJack** [comment]
Aha!!! Found it, to preserve attributes you have to hardlink the files

So, if you want to copy a symlink, you have to hardlink the link to another path, and hardlink will preserve the attributes

**Comment by kegesch** [comment]
Yes, but a hardlink is not a symlink if i understood correctly.

**Comment by GrayJack** [comment]
Yes, but all code that I saw (FreeBSD, NetBSD and uutils) does that when preserving a copy a symlink preserving metadata
